library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/ResourceLoaderImpl.java

```diff
@@ -0,0 +1,325 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.resource.AbstractFileResourcePack;
+import net.minecraft.resource.DefaultResourcePack;
+import net.minecraft.resource.Resource;
+import net.minecraft.resource.ResourceImpl;
+import net.minecraft.resource.ResourcePack;
+import net.minecraft.resource.ResourcePackProfile;
+import net.minecraft.resource.ResourceReloader;
+import net.minecraft.resource.ResourceType;
+import net.minecraft.util.Identifier;
+
+import org.quiltmc.qsl.resource.loader.api.GroupResourcePack;
+import org.quiltmc.qsl.resource.loader.api.ResourceLoader;
+import org.quiltmc.qsl.resource.loader.api.ResourcePackActivationType;
+import org.quiltmc.qsl.resource.loader.api.reloader.IdentifiableResourceReloader;
+import org.quiltmc.qsl.resource.loader.mixin.NamespaceResourceManagerAccessor;
+
+/**
+ * Represents the implementation of the resource loader.
+ */
+@ApiStatus.Internal
+public final class ResourceLoaderImpl implements ResourceLoader {
+	private static final Map<ResourceType, ResourceLoaderImpl> IMPL_MAP = new EnumMap<>(ResourceType.class);
+	private static final Map<String, ModNioResourcePack> CLIENT_BUILTIN_RESOURCE_PACKS = new Object2ObjectOpenHashMap<>();
+	private static final Map<String, ModNioResourcePack> SERVER_BUILTIN_RESOURCE_PACKS = new Object2ObjectOpenHashMap<>();
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	private final Set<Identifier> addedListenerIds = new HashSet<>();
+	private final Set<IdentifiableResourceReloader> addedReloaders = new LinkedHashSet<>();
+
+	public static ResourceLoaderImpl get(ResourceType type) {
+		return IMPL_MAP.computeIfAbsent(type, t -> new ResourceLoaderImpl());
+	}
+
+	/* Resource reloaders stuff */
+
+	public static void sort(ResourceType type, List<ResourceReloader> reloaders) {
+		get(type).sort(reloaders);
+	}
+
+	@Override
+	public void registerReloader(IdentifiableResourceReloader resourceReloader) {
+		if (!this.addedListenerIds.add(resourceReloader.getQuiltId())) {
+			throw new IllegalStateException(
+					"Tried to register resource reloader " + resourceReloader.getQuiltId() + " twice!"
+			);
+		}
+
+		if (!this.addedReloaders.add(resourceReloader)) {
+			throw new IllegalStateException(
+					"Resource reloader with previously unknown ID " + resourceReloader.getQuiltId()
+							+ " already in resource reloader set!"
+			);
+		}
+	}
+
+	/**
+	 * Sorts the given resource reloaders to satisfy dependencies.
+	 *
+	 * @param reloaders the resource reloaders to sort
+	 */
+	private void sort(List<ResourceReloader> reloaders) {
```

Comment:
Would be nice to place comments inside the method dictating each step.

Reply:
Done.